### PR TITLE
drivers ADIS16448 minor fixes and compatibility with older model

### PR DESCRIPTION
**Pull request by dagar**
For debugging https://github.com/PX4/PX4-Autopilot/issues/17509.

### CODE REVIEW

**Comment by Jaeyoung-Lim** [comment]
@dagar Thanks!

This is the output of the branch: 
```
sercon: Registering CDC/ACM serial driver
sercon: Successfully registered the CDC/ACM serial driver
HW arch: PX4_FMU_V5
HW type: V500
HW version: 0x00000000
HW revision: 0x00000000
FW git-hash: 34f72b9f899213511244ab90e63aea2353f4851e
FW version: 1.12.0 80 (17563776)
FW git-branch: pr-adis16448_probe_retry
OS: NuttX
OS version: Release 10.0.0 (167772415)
OS git-hash: 405cfa2571ab9eee2fd8cdf28726f8ccfeb3e5e6
Build datetime: May  2 2021 20:49:57
Build uri: localhost
Toolchain: GNU GCC, 9.3.1 20200408 (release)
PX4GUID: 00020000000039323833343851140045002d
MCU: STM32F76xxx, rev. Z
INFO  [param] selected parameter default file /fs/mtd_params
INFO  [parameters] BSON document size 1167 bytes, decoded 1167 bytes
Board architecture defaults: /etc/init.d/rc.board_arch_defaults
Board defaults: /etc/init.d/rc.board_defaults
INFO  [uavcan] Node ID 1, bitrate 1000000
INFO  [px4io] IO FW CRC match
Board sensors: /etc/init.d/rc.board_sensors
icm20689 #0 on SPI bus 1 (devid=0x3c)
bmi055 #0 on SPI bus 1 (devid=0x41)
bmi055 #1 on SPI bus 1 (devid=0x42)
ms5611 #0 on SPI bus 4 (devid=0x3d)
ist8310 #0 on I2C bus 3
WARN  [SPI_I2C] ist8310: no instance started (no device on bus?)
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 probe failed
adis16448 SPI::init failed (-1)
WARN  [SPI_I2C] adis16448: no instance started (no device on bus?)
Board extras: /etc/init.d/rc.board_mavlink
INFO  [mavlink] mode: Config, data rate: 800000 B/s on /dev/ttyACM0 @ 57600B
Starting Main GPS on /dev/ttyS0
Starting MAVLink on /dev/ttyS1
INFO  [mavlink] mode: Normal, data rate: 1200 B/s on /dev/ttyS1 @ 57600B
No autostart ID found
ekf2 [253:100]
INFO  [logger] logger started (mode=all)
INFO  [logger] Logging FIFO data: increasing task prio and logging rate

NuttShell (NSH) NuttX-10.0.0
nsh> ERROR [ekf2] 0 - vehicle_magnetometer lost, generation 10 -> 12
INFO  [ekf2] starting instance 0, IMU:0 (3932170), MAG:0 (396825)
INFO  [ekf2] starting instance 1, IMU:1 (4259850), MAG:0 (396825)
INFO  [mavlink] Starting mavlink shell
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 probe failed
adis16448 SPI::init failed (-1)
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
INFO  [adis16448] Serial Number: 0xdf, Lot ID1: 0x15 ID2: 0x00
adis16448 on SPI bus 5 at 0 (2000 KHz)
```

**Comment by dagar** [comment]
So no difference?

Does it always start eventually after a few attempts?

**Comment by Jaeyoung-Lim** [comment]
@dagar Yes, but even with 5 retries sometimes it seems to fail.

The driver still is stuck in the configure phase

**Comment by dagar** [comment]
> The driver still is stuck in the configure phase

Can you add `-DDEBUG_BUILD` in the compile options and capture specifically what's failing in the configure?

**Comment by Jaeyoung-Lim** [comment]
@dagar Results with debug enabled:

```
INFO  [mavlink] Starting mavlink shell
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 probe failed
adis16448 SPI::init failed (-1)
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
adis16448 unexpected PROD_ID 0x40
adis16448 unexpected PROD_ID 0x00
INFO  [adis16448] Serial Number: 0xdf, Lot ID1: 0x15 ID2: 0x00
adis16448 on SPI bus 5 at 0 (2000 KHz)
DEBUG 714488286  self test passed
DEBUG 714578477  Reset not complete, check again in 10 ms
DEBUG 714678651  Reset not complete, check again in 10 ms
DEBUG 714778842  Reset not complete, check again in 10 ms
DEBUG 714879044  Reset not complete, check again in 10 ms
DEBUG 714979243  Reset not complete, check again in 10 ms
DEBUG 715079439  Reset not complete, check again in 10 ms
DEBUG 715179593  Reset not complete, check again in 10 ms
DEBUG 715279742  Reset not complete, check again in 10 ms
DEBUG 715379897  Reset not complete, check again in 10 ms
DEBUG 715480040  Reset not complete, check again in 10 ms
DEBUG 715580192  Reset failed, retrying
DEBUG 715770410  Reset not complete, check again in 10 ms
DEBUG 715870557  Reset not complete, check again in 10 ms
DEBUG 715970758  Reset not complete, check again in 10 ms
DEBUG 716070968  Reset not complete, check again in 10 ms
DEBUG 716171114  Reset not complete, check again in 10 ms
DEBUG 716271261  Reset not complete, check again in 10 ms
DEBUG 716371411  Reset not complete, check again in 10 ms
DEBUG 716471569  Reset not complete, check again in 10 ms
DEBUG 716571725  Reset not complete, check again in 10 ms
DEBUG 716671871  Reset not complete, check again in 10 ms
DEBUG 716772017  Reset failed, retrying
DEBUG 716962237  Reset not complete, check again in 10 ms
nsh> dmesg
 check again in 10 ms
DEBUG 717644900  Reset not complete, check again in 10 ms
DEBUG 717745050  Reset not complete, check again in 10 ms
DEBUG 717845211  Reset not complete, check again in 10 ms
DEBUG 717945385  Reset failed, retrying
DEBUG 718135566  Reset not complete, check again in 10 ms
DEBUG 718235753  Reset not complete, check again in 10 ms
DEBUG 718335900  Reset not complete, check again in 10 ms
DEBUG 718436047  Reset not complete, check again in 10 ms
DEBUG 718536201  Reset not complete, check again in 10 ms
DEBUG 718636404  Reset not complete, check again in 10 ms
DEBUG 718736577  Reset not complete, check again in 10 ms
DEBUG 718836729  Reset not complete, check again in 10 ms
DEBUG 718936940  Reset not complete, check again in 10 ms
DEBUG 719037198  Reset not complete, check again in 10 ms
DEBUG 719137467  Reset failed, retrying
DEBUG 719327639  Reset not complete, check again in 10 ms
DEBUG 719427794  Reset not complete, check again in 10 ms
DEBUG 719527940  Reset not complete, check again in 10 ms
DEBUG 719628085  Reset not complete, check again in 10 ms
DEBUG 719728554  0x34: 0x00 (0x10 not set)
DEBUG 719728684  0x36: 0x00 (0x01 not set)
DEBUG 719728800  0x38: 0x00 (0x00 not set)
DEBUG 719729023  Configure failed, retrying
DEBUG 719829401  0x34: 0x00 (0x10 not set)
DEBUG 719829526  0x36: 0x00 (0x01 not set)
DEBUG 719829646  0x38: 0x04 (0x07 not cleared)
DEBUG 719829770  0x32: 0x0C (0x03 not set)
DEBUG 719829861  Configure failed, retrying
DEBUG 719930219  0x34: 0x00 (0x10 not set)
DEBUG 719930350  0x36: 0x00 (0x01 not set)
DEBUG 719930466  0x38: 0x04 (0x00 not set)
DEBUG 719930620  0x38: 0x04 (0x07 not cleared)
DEBUG 719930808  0x32: 0x00 (0x03 not set)
DEBUG 719930901  Configure failed, retrying
DEBUG 720031377  0x34: 0x00 (0x10 not set)
DEBUG 720031541  0x36: 0x00 (0x01 not set)
DEBUG 720031663  0x38: 0x04 (0x00 not set)
DEBUG 720031752  0x38: 0x04 (0x07 not cleared)
DEBUG 720031872  0x32: 0x00 (0x03 not set)
DEBUG 720031958  Configure failed, retrying
DEBUG 720132317  0x34: 0x00 (0x10 not set)
DEBUG 720132447  0x36: 0x00 (0x01 not set)
DEBUG 720132563  0x38: 0x04 (0x00 not set)
DEBUG 720132658  0x38: 0x04 (0x07 not cleared)
DEBUG 720132783  0x32: 0x00 (0x03 not set)
DEBUG 720132871  Configure failed, retrying
DEBUG 720233223  0x34: 0x00 (0x10 not set)
DEBUG 720233477  0x36: 0x00 (0x01 not set)
DEBUG 720233593  0x38: 0x04 (0x00 not set)
DEBUG 720233685  0x38: 0x04 (0x07 not cleared)
DEBUG 720233854  0x32: 0x00 (0x03 not set)
DEBUG 720233952  Configure failed, retrying
DEBUG 720334289  0x34: 0x00 (0x10 not set)
DEBUG 720334421  0x36: 0x00 (0x01 not set)
DEBUG 720334535  0x38: 0x04 (0x00 not set)
DEBUG 720334623  0x38: 0x04 (0x07 not cleared)
DEBUG 720334762  0x32: 0x00 (0x03 not set)
DEBUG 720334852  Configure failed, resetting
DEBUG 720525017  Reset not complete, check again in 10 ms
DEBUG 720625171  Reset not complete, check again in 10 ms
DEBUG 720725353  Reset not complete, check again in 10 ms
DEBUG 720825510  Reset not complete, check again in 10 ms
DEBUG 720925663  Reset not complete, check again in 10 ms
DEBUG 721025816  Reset not complete, check again in 10 ms
DEBUG 721125971  Reset not complete, check again in 10 ms
DEBUG 721226117  Reset not complete, check again in 10 ms
DEBUG 721326264  Reset not complete, check again in 10 ms
DEBUG 721426414  Reset not complete, check again in 10 ms
DEBUG 721526611  Reset failed, retrying
DEBUG 721716829  Reset not complete, check again in 10 ms
DEBUG 721816976  Reset not complete, check again in 10 ms
DEBUG 721917166  Reset not complete, check again in 10 ms
DEBUG 722017322  Reset not complete, check again in 10 ms
DEBUG 722117470  Reset not complete, check again in 10 ms
DEBUG 722217620  Reset not complete, check again in 10 ms
DEBUG 722317773  Reset not complete, check again in 10 ms
DEBUG 722417964  Reset not complete, check again in 10 ms
DEBUG 722518226  Reset not complete, check again in 10 ms
DEBUG 722618376  Reset not complete, check again in 10 ms
DEBUG 722718522  Reset failed, retrying
```

**Comment by Jaeyoung-Lim** [comment]
@dagar I think the readregister is just unreliable, and this is failing not only on starting the driver (unexpected prod ID) but also failing during configuration.

**Review by Jaeyoung-Lim — DISMISSED** [review]
Thank you so much @dagar !

File: src/drivers/imu/analog_devices/adis16448/ADIS16448.cpp [hunk 1]
```diff
@@ -208,23 +207,43 @@ void ADIS16448::RunImpl()
 				if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
 					PX4_DEBUG("Reset failed, retrying");
 					_state = STATE::RESET;
-					ScheduleDelayed(100_ms);
+					ScheduleDelayed(1000_ms);
```
Comment (dagar) [inline src/drivers/imu/analog_devices/adis16448/ADIS16448.cpp:210]:
```suggestion
					ScheduleDelayed(100_ms);
```

**Pull request merged**